### ISSUE TITLE
Intel 11th Gen. JSL: Assertion failed: buf_vk->mem.data, file ../src/vulkan/gpu_buf.c, line 260

### ISSUE THREAD

**Issue by Chipcraft** [haasn/libplacebo#146]
- Intel Pentium Silver N6000 (Jasper Lake), Intel UHD 32EU w/ 31.0.101.2114 drivers
- Windows 10 Pro 19045.2130 22H2

Configure:

$ meson -Dbuildtype=release -Dtests=true -Dbench=true Build
The Meson build system
Version: 0.63.3
Source dir: C:/msys64/home/Poelloe_Z3/libplacebo
Build dir: C:/msys64/home/Poelloe_Z3/libplacebo/Build
Build type: native build
Project name: libplacebo
Project version: 5.228.0
C compiler for the host machine: cc (gcc 12.2.0 "cc (Rev4, Built by MSYS2 project) 12.2.0")
C linker for the host machine: cc ld.bfd 2.39
C++ compiler for the host machine: c++ (gcc 12.2.0 "c++ (Rev4, Built by MSYS2 project) 12.2.0")
C++ linker for the host machine: c++ ld.bfd 2.39
Host machine cpu family: x86_64
Host machine cpu: x86_64
Compiler for C supports arguments -Wincompatible-pointer-types: YES
Compiler for C supports link arguments -Wl,--exclude-libs=ALL: YES
Library m found: YES
Program python3 found: YES (C:/msys64/mingw64/bin/python.exe)
Found pkg-config: C:\msys64\mingw64\bin/pkg-config.EXE (1.8.0)
Found CMake: C:\msys64\mingw64\bin/cmake.EXE (3.24.2)
Run-time dependency libunwind found: NO (tried pkgconfig and cmake)
Library execinfo found: NO
Checking for function "backtrace_symbols" with dependency -lexecinfo: NO
Run-time dependency shaderc found: YES 2022.3.1
Message: Skipping `glslang` because `shaderc` is available
Check usable header "d3d11.h" : YES
Check usable header "d3d11_4.h" : YES
Check usable header "dxgi1_6.h" : YES
Run-time dependency spirv-cross-c-shared found: YES 0.49.0
Library version found: YES
Run-time dependency vulkan found: YES 1.3.231
Run-time dependency lcms2 found: YES 2.13
Configuring config.h using configuration
Configuring config_internal.h using configuration
Run-time dependency dav1d found: NO (tried pkgconfig and cmake)
Run-time dependency libavutil found: NO (tried pkgconfig and cmake)
Run-time dependency libavcodec found: NO (tried pkgconfig and cmake)
Run-time dependency libavformat found: NO (tried pkgconfig and cmake)
Configuring include_colorspace_h.c using configuration
Configuring include_common_h.c using configuration
Configuring include_d3d11_h.c using configuration
Configuring include_dispatch_h.c using configuration
Configuring include_dither_h.c using configuration
Configuring include_dummy_h.c using configuration
Configuring include_filters_h.c using configuration
Configuring include_gpu_h.c using configuration
Configuring include_log_h.c using configuration
Configuring include_opengl_h.c using configuration
Configuring include_renderer_h.c using configuration
Configuring include_shaders_colorspace_h.c using configuration
Configuring include_shaders_custom_h.c using configuration
Configuring include_shaders_deinterlacing_h.c using configuration
Configuring include_shaders_dithering_h.c using configuration
Configuring include_shaders_film_grain_h.c using configuration
Configuring include_shaders_icc_h.c using configuration
Configuring include_shaders_lut_h.c using configuration
Configuring include_shaders_sampling_h.c using configuration
Configuring include_shaders_h.c using configuration
Configuring include_swapchain_h.c using configuration
Configuring include_tone_mapping_h.c using configuration
Configuring include_utils_frame_queue_h.c using configuration
Configuring include_utils_upload_h.c using configuration
Configuring include_vulkan_h.c using configuration
Run-time dependency glfw3 found: NO (tried pkgconfig and cmake)
sdl2-config found: NO
Run-time dependency sdl2 found: NO (tried pkgconfig and config-tool)
Run-time dependency sdl2_image found: NO (tried pkgconfig and cmake)
Run-time dependency threads found: YES
Run-time dependency libavcodec found: NO (tried pkgconfig and cmake)
Run-time dependency libavformat found: NO (tried pkgconfig and cmake)
Run-time dependency libavutil found: NO (tried pkgconfig and cmake)
Has header "nuklear.h" : YES
Configuring config_demos.h using configuration
demos/meson.build:97: WARNING: Demos enabled but no supported combination of windowing system and gr
aphical APIs was found. Demo programs require either GLFW or SDL and either Vulkan or OpenGL to func
tion.
Build targets in project: 45

libplacebo 5.228.0

  Optional features
    d3d11       : YES
    gl-proc-addr: YES
    glslang     : NO
    lcms        : YES
    opengl      : YES
    shaderc     : YES
    vk-proc-addr: YES
    vulkan      : YES

  User defined options
    buildtype   : release
    bench       : true
    tests       : true

Found ninja-1.11.1 at C:\msys64\mingw64\bin/ninja.EXE

---------------

Build: 

$ ninja -CBuild
ninja: Entering directory `Build'
[44/146] Compiling C object src/libplacebo-228.dll.p/d3d11_utils.c.obj
../src/d3d11/utils.c: In function 'pl_get_dxgi_csp_name':
../src/d3d11/utils.c:316:43: warning: case value '20' not in enumerated type 'DXGI_COLOR_SPACE_TYPE'
 [-Wswitch]
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: note: in definition of macro 'D3D11_DXGI_ENUM'
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: warning: case value '21' not in enumerated type 'DXGI_COLOR_SPACE_TYPE'
 [-Wswitch]
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: note: in definition of macro 'D3D11_DXGI_ENUM'
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: warning: case value '22' not in enumerated type 'DXGI_COLOR_SPACE_TYPE'
 [-Wswitch]
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: note: in definition of macro 'D3D11_DXGI_ENUM'
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: warning: case value '23' not in enumerated type 'DXGI_COLOR_SPACE_TYPE'
 [-Wswitch]
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: note: in definition of macro 'D3D11_DXGI_ENUM'
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: warning: case value '24' not in enumerated type 'DXGI_COLOR_SPACE_TYPE'
 [-Wswitch]
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
../src/d3d11/utils.c:316:43: note: in definition of macro 'D3D11_DXGI_ENUM'
  316 | #define D3D11_DXGI_ENUM(prefix, define) { case prefix ## define: return #define; }
      |                                           ^~~~
[146/146] Linking target src/bench.exe

---------------

Test:

$ ninja -CBuild test
ninja: Entering directory `Build'
 1/14 colorspace.c                 OK               0.45s
 2/14 common.c                     OK               0.44s
 3/14 dither.c                     OK               0.56s
 4/14 filters.c                    OK               0.09s
 5/14 lut.c                        OK               0.27s
 6/14 string.c                     OK               0.16s
 7/14 tone_mapping.c               OK               0.12s
 8/14 dummy.c                      OK               0.72s
 9/14 utils.c                      OK               0.14s
10/14 opengl_surfaceless.c         SKIP             0.12s   exit status 77
11/14 icc.c                        OK               0.06s
12/14 vulkan.c                     FAIL             5.75s   exit status 3
>>> MALLOC_PERTURB_=42 PATH=C:/msys64/home/Poelloe_Z3/libplacebo/Build/src;C:\msys64\mingw64\bin;C:\
msys64\usr\local\bin;C:\msys64\usr\bin;C:\msys64\usr\bin;C:\Windows\System32;C:\Windows;C:\Windows\S
ystem32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\msys64\usr\bin\site_perl;C:\msys64\usr\b
in\vendor_perl;C:\msys64\usr\bin\core_perl;C:\msys64\mingw64\bin\ C:/msys64/home/Poelloe_Z3/libplace
bo/Build/src/tests/test.vulkan.c.exe
――――――――――――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――――――――――――
Listing only the last 100 lines from a long log.
[5.577][d] [ 15] _tex_coord_1_0 = _tex_coord_1_0_v;
[5.577][d] [ 16] vec2 va_pos = _position_5_0_v;
[5.577][d] [ 17] gl_Position = vec4(va_pos, 0.0, 1.0);
[5.577][d] [ 18] }
[5.577][d] fragment shader source:
[5.577][d] [  1] #version 450
[5.577][d] [  2] #extension GL_KHR_shader_subgroup_basic : enable
[5.577][d] [  3] #extension GL_KHR_shader_subgroup_vote : enable
[5.577][d] [  4] #extension GL_KHR_shader_subgroup_arithmetic : enable
[5.577][d] [  5] #extension GL_KHR_shader_subgroup_ballot : enable
[5.577][d] [  6] #extension GL_KHR_shader_subgroup_shuffle : enable
[5.577][d] [  7] #extension GL_ARB_texture_gather : enable
[5.577][d] [  8] layout(constant_id=0) const float _const_2_0 = 1;
[5.577][d] [  9] layout(constant_id=1) const float _const_3_0 = 1;
[5.577][d] [ 10] layout(binding=0) uniform  sampler2D _src_tex_0_0;
[5.577][d] [ 11] layout(location=0) in vec2 _tex_coord_1_0;
[5.577][d] [ 12] layout(location=0)  out vec4 out_color;
[5.577][d] [ 13]
[5.577][d] [ 14] vec4 _main_6_0() {
[5.577][d] [ 15] // pl_shader_sample_direct
[5.577][d] [ 16] vec4 color = vec4(_const_2_0) * texture(_src_tex_0_0, _tex_coord_1_0);
[5.577][d] [ 17] color *= vec4(1.0 / _const_3_0);
[5.577][d] [ 18] vec4 _orig_color_4_0 = color;
[5.577][d] [ 19] color = vec4(0.0, 0.0, 0.0, 1.0);
[5.577][d] [ 20] color[0] = _orig_color_4_0[1];
[5.577][d] [ 21] return color;
[5.577][d] [ 22] }
[5.577][d] [ 23]
[5.577][d] [ 24] void main() {
[5.577][d] [ 25] out_color = _main_6_0();
[5.577][d] [ 26] }
[5.577][d] Specialization constant values:
[5.577][d]   constant_id=0: 1.000000
[5.577][d]   constant_id=1: 1.000000
[5.593][d] shaderc compile status 'success' (0 errors, 0 warnings)
[5.593][i] Spent 16.000 ms translating SPIR-V
[5.593][d] shaderc compile status 'success' (0 errors, 0 warnings)
[5.593][d] Spent 0.000 ms translating SPIR-V
[5.593][d] Spent 0.000 ms compiling shader
[5.593][d] Spent 0.000 ms creating pipeline
[5.593][d] Pass statistics: size 4636, SPIR-V: vert 748 frag 760 comp 0
[5.593][d] vertex shader source:
[5.593][d] [  1] #version 450
[5.593][d] [  2] #extension GL_KHR_shader_subgroup_basic : enable
[5.593][d] [  3] #extension GL_KHR_shader_subgroup_vote : enable
[5.593][d] [  4] #extension GL_KHR_shader_subgroup_arithmetic : enable
[5.593][d] [  5] #extension GL_KHR_shader_subgroup_ballot : enable
[5.593][d] [  6] #extension GL_KHR_shader_subgroup_shuffle : enable
[5.593][d] [  7] #extension GL_ARB_texture_gather : enable
[5.593][d] [  8] layout(constant_id=0) const float _const_2_0 = 1;
[5.593][d] [  9] layout(constant_id=1) const float _const_3_0 = 1;
[5.593][d] [ 10] layout(binding=0) uniform  sampler2D _src_tex_0_0;
[5.593][d] [ 11] layout(location=0) in vec2 _tex_coord_1_0_v;
[5.593][d] [ 12] layout(location=0) out vec2 _tex_coord_1_0;
[5.593][d] [ 13] layout(location=1) in vec2 _position_5_0_v;
[5.593][d] [ 14] void main() {
[5.593][d] [ 15] _tex_coord_1_0 = _tex_coord_1_0_v;
[5.593][d] [ 16] vec2 va_pos = _position_5_0_v;
[5.593][d] [ 17] gl_Position = vec4(va_pos, 0.0, 1.0);
[5.593][d] [ 18] }
[5.593][d] fragment shader source:
[5.593][d] [  1] #version 450
[5.593][d] [  2] #extension GL_KHR_shader_subgroup_basic : enable
[5.593][d] [  3] #extension GL_KHR_shader_subgroup_vote : enable
[5.593][d] [  4] #extension GL_KHR_shader_subgroup_arithmetic : enable
[5.593][d] [  5] #extension GL_KHR_shader_subgroup_ballot : enable
[5.593][d] [  6] #extension GL_KHR_shader_subgroup_shuffle : enable
[5.593][d] [  7] #extension GL_ARB_texture_gather : enable
[5.593][d] [  8] layout(constant_id=0) const float _const_2_0 = 1;
[5.593][d] [  9] layout(constant_id=1) const float _const_3_0 = 1;
[5.593][d] [ 10] layout(binding=0) uniform  sampler2D _src_tex_0_0;
[5.593][d] [ 11] layout(location=0) in vec2 _tex_coord_1_0;
[5.593][d] [ 12] layout(location=0)  out vec4 out_color;
[5.593][d] [ 13]
[5.593][d] [ 14] vec4 _main_6_0() {
[5.593][d] [ 15] // pl_shader_sample_direct
[5.593][d] [ 16] vec4 color = vec4(_const_2_0) * texture(_src_tex_0_0, _tex_coord_1_0);
[5.593][d] [ 17] color *= vec4(1.0 / _const_3_0);
[5.593][d] [ 18] vec4 _orig_color_4_0 = color;
[5.593][d] [ 19] color = vec4(0.0, 0.0, 0.0, 1.0);
[5.593][d] [ 20] color[0] = _orig_color_4_0[2];
[5.593][d] [ 21] return color;
[5.593][d] [ 22] }
[5.593][d] [ 23]
[5.593][d] [ 24] void main() {
[5.593][d] [ 25] out_color = _main_6_0();
[5.593][d] [ 26] }
[5.593][d] Specialization constant values:
[5.593][d]   constant_id=0: 1.000000
[5.593][d]   constant_id=1: 1.000000
[5.608][d] shaderc compile status 'success' (0 errors, 0 warnings)
[5.608][i] Spent 15.000 ms translating SPIR-V
[5.608][d] shaderc compile status 'success' (0 errors, 0 warnings)
[5.608][d] Spent 0.000 ms translating SPIR-V
[5.608][d] Spent 0.000 ms compiling shader
[5.608][d] Spent 0.000 ms creating pipeline
[5.608][d] Pass statistics: size 4700, SPIR-V: vert 748 frag 760 comp 0
stderr:
[0.000][w] API debugging requested but no debug meta layers present... ignoring
Assertion failed: buf_vk->mem.data, file ../src/vulkan/gpu_buf.c, line 260
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

13/14 d3d11.c                      OK              23.20s
14/14 benchmark                    FAIL             0.97s   exit status 3
>>> MALLOC_PERTURB_=62 PATH=C:/msys64/home/Poelloe_Z3/libplacebo/Build/src;C:\msys64\mingw64\bin;C:\
msys64\usr\local\bin;C:\msys64\usr\bin;C:\msys64\usr\bin;C:\Windows\System32;C:\Windows;C:\Windows\S
ystem32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\msys64\usr\bin\site_perl;C:\msys64\usr\b
in\vendor_perl;C:\msys64\usr\bin\core_perl;C:\msys64\mingw64\bin\ C:/msys64/home/Poelloe_Z3/libplace
bo/Build/src/bench.exe
――――――――――――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――――――――――――
stdout:
= Running benchmarks =
stderr:
Assertion failed: buf_vk->mem.data, file ../src/vulkan/gpu_buf.c, line 260
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

Summary of Failures:

12/14 vulkan.c             FAIL             5.75s   exit status 3
14/14 benchmark            FAIL             0.97s   exit status 3

Ok:                 11
Expected Fail:      0
Fail:               2
Unexpected Pass:    0
Skipped:            1
Timeout:            0

Full log written to C:/msys64/home/Poelloe_Z3/libplacebo/Build/meson-logs/testlog.txt

FAILED: meson-internal__test
"C:/msys64/mingw64/bin/meson" "test" "--no-rebuild" "--print-errorlogs"
ninja: build stopped: subcommand failed.

Testlog.txt as attachment.

[testlog.zip](https://github.com/haasn/libplacebo/files/9863388/testlog.zip)

Intel drivers are a total shit show and I fully expect them to be the cause of the failure however, I'm interested if there is anything that can be done / tried from the software side?

**Comment by haasn** [comment]
Do you reckon it would be possible for you to point a `gdb` at the asserting process and throw a `p *buf` and `p *buf_vk` my way?

**Comment by Chipcraft** [comment]
I have no experience with the debuggers however, while trying to run "test.vulkan.c.exe" attached to gdb I noticed something:
The very same binary which crashes when executed either manually (stand-alone) or through ninja works perfectly fine when executed with the debugger attached.

It looks to be some kind of a timing / timeout issue, related to the dynamic graphics memory allocation (DVMT). When the binaries are built with O0 optimization level (-Dc_args="-O0 -g" -Dcpp_args="-O0 -g") the resulting binaries are slow enough to work. If the binaries are built with e.g., >= O1 optimization level, they fail unless they're executed with the debugger attached, which on a system this slow will slow down to a crawl.

So, to summarize:

The default config: meson -Dtests=true -Dbenchmark=true >> optimization level is O2 w/ -g, crashes unless executed with gdb attached
- No-opt. config: meson -Dtests=true -Dbenchmark=true -Dc_args="-O0 -g" -Dcpp_args="-O0 -g" >> works fine either standalone or with gdb attached
- O1 opt. config: -Dtests=true -Dbenchmark=true -Dc_args="-O1 -g" -Dcpp_args="-O1 -g" >> crashes, unless executed with gdb attached

Attached the output of a successful stand-alone "O0" run, not that I'd expect it to make much of a difference.
[O0_Success.txt](https://github.com/haasn/libplacebo/files/9864749/O0_Success.txt)

**Comment by Chipcraft** [comment]
Not sure if this GDB output is what you asked for, but in any case:

[gpu_buf_260_break.txt](https://github.com/haasn/libplacebo/files/9865361/gpu_buf_260_break.txt)

**Comment by haasn** [comment]
Thanks, that does look interesting. It seems like what's happening is that this is a host pointer import, but for some reason, it gets imported as a memory type that is not itself host-visible (i.e. cannot be mapped back to RAM). This is pretty much fine for our purposes, but it does mean that the assumption that all host-imported pointers are also host-mapped is wrong.

This patch *should* fix it. Can you test?
```diff
diff --git a/src/vulkan/gpu_buf.c b/src/vulkan/gpu_buf.c
index 037235d7..d1cf837f 100644
--- a/src/vulkan/gpu_buf.c
+++ b/src/vulkan/gpu_buf.c
@@ -256,8 +256,7 @@ static void invalidate_buf(pl_gpu gpu, pl_buf buf)
     struct vk_ctx *vk = p->vk;
     struct pl_buf_vk *buf_vk = PL_PRIV(buf);
 
-    if (!buf_vk->mem.coherent) {
-        pl_assert(buf_vk->mem.data);
+    if (buf_vk->mem.data && !buf_vk->mem.coherent) {
         VK(vk->InvalidateMappedMemoryRanges(vk->dev, 1, &(VkMappedMemoryRange) {
             .sType = VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE,
             .memory = buf_vk->mem.vkmem,
```

That said, I have no idea how or why the choice of debugging settings would affect this outcome in any way...

**Comment by Chipcraft** [comment]
With the patch in place, the tests and the benchmark complete regardless of the optimization level / execution speed.

The testlog however shows "error: vk->BeginCommandBuffer(cmd->buf, &binfo): VK_ERROR_OUT_OF_HOST_MEMORY (../src/vulkan/command.c:325)" being spammed, again probably due to the dynamic GPU DRAM allocation (DVMT). The iGPU is allowed to use half of the available system DRAM (32GB) however, there is no way to set a fixed / guaranteed allocation for it.

Likewise, FFMpeg HWUpload is failing with memory related error, when using a command which works on AMD* and NVIDIA GPUs:

[AVHWFramesContext @ 00000205dbdfa300] Cannot allocate all planes in a single allocation, device requires dedicated image allocation!
[hwupload @ 00000205dc68f780] Failed to allocate frame to upload to.
Error while filtering: Cannot allocate memory
Failed to inject frame into filter network: Cannot allocate memory
Error while processing the decoded data for stream #0:0
Conversion failed!

So, with AMD GPUs (under Windows, [)](https://github.com/haasn/libplacebo/issues/127) being unusable due to the driver issues and Intel iGPUs being unusable, at least for the most part due to the shared memory situation, libplacebo currently is pretty much for NVIDIA only. 

Testlog as attachment.
[testlog.zip](https://github.com/haasn/libplacebo/files/9871868/testlog.zip)

I guess we can close this, unless there is something further I could try out?

**Comment by haasn** [comment]
That testlog only shows the errors for the `benchmark`, which is designed to test your GPU's throughput. It's possible that the volume of commands we're submitting *actually* exceeds your system's memory capabilities. (Though to be honest, I would find that shocking, and it's more likely there's a driver bug)

If it's only the benchmark showing those errors, and everything else works, isn't that more or less fine?

> Likewise, FFMpeg HWUpload is failing with memory related error, when using a command which works on AMD* and NVIDIA GPUs:

FFmpeg bug, can you please report it there?
